components/core/src/package/archive.rs

```diff
@@ -417,6 +419,77 @@ pub trait FromArchive: Sized {
     fn from_archive(archive: &mut PackageArchive) -> result::Result<Self, Self::Error>;
 }
 
+// Exposes the extended archive and header metadata
+#[derive(Deserialize, Serialize)]
+pub struct PackageArchiveInfo {
+    pub format_version:    String,
+    pub key_name:          String,
+    pub hash_type:         String,
+    pub signature_raw:     String,
+    pub origin:            String,
+    pub name:              String,
+    pub version:           Option<String>,
+    pub release:           Option<String>,
```

Comment:
It would be really nice to include a key for the complete identifier all in one string (e.g., `core/xz/5.2.4/20190115013348`)

When I'm doing some automation around Habitat packages using `jq`, I _reeeeeeeeally_ hate doing

```
x.origin + "/" + x.name + "/" + x.version + "/" x.release
```
all... the... time.

Reply:
Totally agree with this. The new output looks like:

```ubuntu@ip-172-31-43-73:~/Projects/habitat$ ./target/debug/hab pkg info ~/.hab/cache/artifacts/habitat-builder-api-8865-20200417184721-x86_64-linux.hart --json | jq
{
  "format_version": "HART-1",
  "key_name": "habitat-20180129222536",
  "hash_type": "BLAKE2b",
  "signature_raw": "jp3QUbh64mMMdCn+7gK0uUVFQj9yWmsh0zgPgsZnBy9TiglsT5md+hmbitinIWC6aLal0x41c57IAD3Cx+9cCDA2OTg2NTcxNjBlYmMwN2YwNmNlYTZlMDA4YzJiZGE2NGUyYzM3ODcxMDhjZWMwMTBjZGIxOGFlODVmOTcxMDk=",
  "origin": "habitat",
  "name": "builder-api",
  "version": "8865",
  "release": "20200417184721",
  "checksum": "84d6bfe32cff6efb0efa1082ec8ad540fcce7c0a7ccdf5a51ab2c5fe938e0fa2",
  "target": "x86_64-linux",
  "is_a_service": true,
  "deps": [
    "core/coreutils/8.30/20190115012313",
    "core/curl/7.68.0/20200309012427",
    "core/gcc-libs/8.2.0/20190115011926",
    "core/glibc/2.27/20190115002733",
    "core/libarchive/3.3.3/20190305214120",
    "core/libsodium/1.0.16/20190116014025",
    "core/openssl/1.0.2r/20190305210149",
    "core/postgresql/9.6.11/20191119235705",
    "core/zeromq/4.3.1/20190802173651"
  ],
  "tdeps": [
    "core/acl/2.2.53/20190115012136",
    "core/attr/2.4.48/20190115012129",
    "core/bash/4.4.19/20190115012619",
    "core/bzip2/1.0.6/20190115011950",
    "core/cacerts/2018.12.05/20190115014206",
    "core/coreutils/8.30/20190115012313",
    "core/curl/7.68.0/20200309012427",
    "core/db/5.3.28/20190115012845",
    "core/gcc-libs/8.2.0/20190115011926",
    "core/gdal/2.2.1/20190115172005",
    "core/gdbm/1.17/20190115012826",
    "core/geos/3.7.1/20190205061756",
    "core/glibc/2.27/20190115002733",
    "core/gmp/6.1.2/20190115003943",
    "core/less/530/20190115013008",
    "core/libarchive/3.3.3/20190305214120",
    "core/libcap/2.25/20190115012150",
    "core/libossp-uuid/1.6.2/20190115171615",
    "core/libsodium/1.0.16/20190116014025",
    "core/libxml2/2.9.10/20191119234603",
    "core/linux-headers/4.17.12/20190115002705",
    "core/ncurses/6.1/20190115012027",
    "core/nghttp2/1.34.0/20190115160823",
    "core/openssl-fips/2.0.16/20190115014207",
    "core/openssl/1.0.2r/20190305210149",
    "core/pcre/8.42/20190115012526",
    "core/perl/5.28.0/20190115013014",
    "core/postgresql/9.6.11/20191119235705",
    "core/proj/4.9.3/20190115171931",
    "core/readline/7.0.3/20190115012607",
    "core/xz/5.2.4/20190115013348",
    "core/zeromq/4.3.1/20190802173651",
    "core/zlib/1.2.11/20190115003728"
  ],
  "build_deps": [
    "core/cacerts/2018.12.05/20190115014206",
    "core/coreutils/8.30/20190115012313",
    "core/gcc/8.2.0/20190115004042",
    "core/git/2.25.1/20200309023931",
    "core/pkg-config/0.29.2/20190115011955",
    "core/protobuf-cpp/3.9.2/20191001233803",
    "core/protobuf-rust/1.7.4/20190116233225",
    "core/rust/1.41.0/20200217103343"
  ],
  "build_tdeps": [
    "core/acl/2.2.53/20190115012136",
    "core/attr/2.4.48/20190115012129",
    "core/binutils/2.31.1/20190115003743",
    "core/busybox-static/1.29.2/20190115014552",
    "core/bzip2/1.0.6/20190115011950",
    "core/cacerts/2018.12.05/20190115014206",
    "core/coreutils/8.30/20190115012313",
    "core/curl/7.68.0/20200309012427",
    "core/db/5.3.28/20190115012845",
    "core/expat/2.2.5/20190115012836",
    "core/gcc-libs/8.2.0/20190115011926",
    "core/gcc/8.2.0/20190115004042",
    "core/gdbm/1.17/20190115012826",
    "core/gettext/0.19.8/20190115013412",
    "core/git/2.25.1/20200309023931",
    "core/glibc/2.27/20190115002733",
    "core/gmp/6.1.2/20190115003943",
    "core/less/530/20190115013008",
    "core/libcap/2.25/20190115012150",
    "core/libmpc/1.1.0/20190115004027",
    "core/linux-headers/4.17.12/20190115002705",
    "core/mpfr/4.0.1/20190115004008",
    "core/ncurses/6.1/20190115012027",
    "core/nghttp2/1.34.0/20190115160823",
    "core/openssh/7.5p1/20190305213650",
    "core/openssl-fips/2.0.16/20190115014207",
    "core/openssl/1.0.2r/20190305210149",
    "core/pcre/8.42/20190115012526",
    "core/perl/5.28.0/20190115013014",
    "core/pkg-config/0.29.2/20190115011955",
    "core/protobuf-cpp/3.9.2/20191001233803",
    "core/protobuf-rust/1.7.4/20190116233225",
    "core/rust/1.41.0/20200217103343",
    "core/sed/4.5/20190115012152",
    "core/xz/5.2.4/20190115013348",
    "core/zlib/1.2.11/20190115003728"
  ],
  "exposes": [
    9636
  ],
  "manifest": "# habitat / builder-api\n\n\n* __Maintainer__: The Habitat Maintainers <humans@habitat.sh>\n* __Version__: 8865\n* __Release__: 20200417184721\n* __Target__: x86_64-linux\n* __Upstream URL__: upstream project's website or home page is not defined\n* __License__: Apache-2.0 \n* __Source__: source URL not provided or required\n* __SHA__: SHA256 checksum not provided or required\n* __Path__: `/hab/pkgs/habitat/builder-api/8865/20200417184721`\n* __Build Dependencies__: `core/protobuf-cpp core/protobuf-rust core/coreutils core/cacerts core/rust core/gcc core/git core/pkg-config `\n* __Dependencies__: `core/glibc core/openssl core/coreutils core/gcc-libs core/zeromq core/libsodium core/libarchive core/curl core/postgresql `\n* __Interpreters__: no interpreters or undefined\n\n# Plan\n\n## Build Flags\n\n```bash\nCFLAGS: -I/hab/pkgs/core/protobuf-cpp/3.9.2/20191001233803/include -I/hab/pkgs/core/gcc/8.2.0/20190115004042/include -I/hab/pkgs/core/glibc/2.27/20190115002733/include -I/hab/pkgs/core/openssl/1.0.2r/20190305210149/include -I/hab/pkgs/core/zeromq/4.3.1/20190802173651/include -I/hab/pkgs/core/libsodium/1.0.16/20190116014025/include -I/hab/pkgs/core/libarchive/3.3.3/20190305214120/include -I/hab/pkgs/core/curl/7.68.0/20200309012427/include -I/hab/pkgs/core/postgresql/9.6.11/20191119235705/include\nCPPFLAGS: -I/hab/pkgs/core/protobuf-cpp/3.9.2/20191001233803/include -I/hab/pkgs/core/gcc/8.2.0/20190115004042/include -I/hab/pkgs/core/glibc/2.27/20190115002733/include -I/hab/pkgs/core/openssl/1.0.2r/20190305210149/include -I/hab/pkgs/core/zeromq/4.3.1/20190802173651/include -I/hab/pkgs/core/libsodium/1.0.16/20190116014025/include -I/hab/pkgs/core/libarchive/3.3.3/20190305214120/include -I/hab/pkgs/core/curl/7.68.0/20200309012427/include -I/hab/pkgs/core/postgresql/9.6.11/20191119235705/include\nCXXFLAGS: -I/hab/pkgs/core/protobuf-cpp/3.9.2/20191001233803/include -I/hab/pkgs/core/gcc/8.2.0/20190115004042/include -I/hab/pkgs/core/glibc/2.27/20190115002733/include -I/hab/pkgs/core/openssl/1.0.2r/20190305210149/include -I/hab/pkgs/core/zeromq/4.3.1/20190802173651/include -I/hab/pkgs/core/libsodium/1.0.16/20190116014025/include -I/hab/pkgs/core/libarchive/3.3.3/20190305214120/include -I/hab/pkgs/core/curl/7.68.0/20200309012427/include -I/hab/pkgs/core/postgresql/9.6.11/20191119235705/include\nLDFLAGS: -L/hab/pkgs/core/protobuf-cpp/3.9.2/20191001233803/lib -L/hab/pkgs/core/rust/1.41.0/20200217103343/lib -L/hab/pkgs/core/gcc/8.2.0/20190115004042/lib -L/hab/pkgs/core/git/2.25.1/20200309023931/lib -L/hab/pkgs/core/glibc/2.27/20190115002733/lib -L/hab/pkgs/core/openssl/1.0.2r/20190305210149/lib -L/hab/pkgs/core/gcc-libs/8.2.0/20190115011926/lib -L/hab/pkgs/core/zeromq/4.3.1/20190802173651/lib -L/hab/pkgs/core/libsodium/1.0.16/20190116014025/lib -L/hab/pkgs/core/libarchive/3.3.3/20190305214120/lib -L/hab/pkgs/core/curl/7.68.0/20200309012427/lib -L/hab/pkgs/core/postgresql/9.6.11/20191119235705/lib\nLD_RUN_PATH: /hab/pkgs/core/glibc/2.27/20190115002733/lib:/hab/pkgs/core/openssl/1.0.2r/20190305210149/lib:/hab/pkgs/core/gcc-libs/8.2.0/20190115011926/lib:/hab/pkgs/core/zeromq/4.3.1/20190802173651/lib:/hab/pkgs/core/libsodium/1.0.16/20190116014025/lib:/hab/pkgs/core/libarchive/3.3.3/20190305214120/lib:/hab/pkgs/core/curl/7.68.0/20200309012427/lib:/hab/pkgs/core/postgresql/9.6.11/20191119235705/lib\n```\n\n## Plan Source\n\n```bash\nsource \"../../../support/ci/builder-base-plan.sh\"\npkg_name=builder-api\npkg_origin=habitat\npkg_maintainer=\"The Habitat Maintainers <humans@habitat.sh>\"\npkg_license=('Apache-2.0')\npkg_bin_dirs=(bin)\npkg_deps=(core/glibc core/openssl core/coreutils core/gcc-libs core/zeromq core/libsodium\ncore/libarchive core/curl core/postgresql)\npkg_build_deps=(core/protobuf-cpp core/protobuf-rust core/coreutils core/cacerts\ncore/rust core/gcc core/git core/pkg-config)\npkg_exports=(\n  [port]=http.port\n)\npkg_exposes=(port)\npkg_binds=(\n  [memcached]=\"port\"\n)\npkg_binds_optional=(\n  [jobsrv]=\"rpc_port\"\n)\nbin=\"bldr-api\"\n```",
  "config": "log_level = \"info\"\n\n[api]\nfeatures_enabled = \"jobsrv\"\ntargets = [\"x86_64-linux\", \"x86_64-linux-kernel2\", \"x86_64-windows\"]\nbuild_targets = [\"x86_64-linux\", \"x86_64-linux-kernel2\", \"x86_64-windows\"]\nbuild_on_upload = true\n\n[http]\nlisten = \"0.0.0.0\"\nport   = 9636\n\n[oauth]\nprovider       = \"github\"\ntoken_url      = \"https://github.com/login/oauth/access_token\"\nuserinfo_url   = \"https://api.github.com/user\"\nredirect_url   = \"\"\nclient_id      = \"\"\nclient_secret  = \"\"\n\n[github]\napi_url        = \"https://api.github.com\"\napp_id         = 5565\nwebhook_secret = \"\"\n\n[s3]\nbackend = \"minio\"\nkey_id = \"depot\"\nsecret_key = \"password\"\nendpoint = \"http://localhost:9000\"\nbucket_name = \"habitat-builder-artifact-store.default\"\n\n[artifactory]\napi_url = \"http://localhost:8080\"\napi_key = \"key\"\nrepo = \"habitat-builder-artifact-store.default\"\n\n[memcache]\nttl = 15\n\n[datastore]\nuser = \"hab\"\npassword = \"\"\ndatabase = \"builder\"\nconnection_retry_ms = 300\nconnection_timeout_sec = 3600\ndb_workers = 4\nhost = \"127.0.0.1\"\nport = 5432",
  "svc_user": "hab",
  "ld_run_path": null,
  "ldflags": null,
  "cflags": null
}
ubuntu@ip-172-31-43-73:~/Projects/habitat$```

Reply:
Looks _much_ nicer!